pkg/wipeout/wipeout.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wipeout
+
+import (
+	"context"
+
+	"github.com/apigee/registry/cmd/registry/core"
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/log"
+	"github.com/apigee/registry/rpc"
+)
+
+// Set this global variable in calling code to run verbosely.
+var Verbose bool
+
+// This collects one (arbitrary) error that occurs during wipeout.
+// It's a global and gets assigned from concurrent goroutines,
+// but if it fails, it fails.
+var _err error
+
+// Wipeout deletes all resources in a project using the specified number of parallel worker jobs.
+func Wipeout(ctx context.Context, client connection.Client, projectid string, jobs int) error {
+	if Verbose {
+		log.Infof(ctx, "Deleting everything in project %s", projectid)
+	}
+	project := "projects/" + projectid + "/locations/global"
+	// Wipeout resources in groups to ensure that children are deleted before parents.
+	{
+		if Verbose {
+			log.Infof(ctx, "Deleting artifacts")
+		}
+		taskQueue, wait := core.WorkerPool(ctx, jobs)
+		wipeoutArtifacts(ctx, client, taskQueue, project+"/apis/-/versions/-/specs/-")
+		wipeoutArtifacts(ctx, client, taskQueue, project+"/apis/-/versions/-")
+		wipeoutArtifacts(ctx, client, taskQueue, project+"/apis/-/deployments/-")
+		wipeoutArtifacts(ctx, client, taskQueue, project+"/apis/-")
+		wipeoutArtifacts(ctx, client, taskQueue, project)
+		wait()
+	}
+	{
+		if Verbose {
+			log.Infof(ctx, "Deleting specs")
+		}
+		taskQueue, wait := core.WorkerPool(ctx, jobs)
+		wipeoutApiSpecs(ctx, client, taskQueue, project+"/apis/-/versions/-")
+		wait()
+	}
+	{
+		if Verbose {
+			log.Infof(ctx, "Deleting versions")
+		}
+		taskQueue, wait := core.WorkerPool(ctx, jobs)
+		wipeoutApiVersions(ctx, client, taskQueue, project+"/apis/-")
+		wait()
+	}
+	{
+		if Verbose {
+			log.Infof(ctx, "Deleting deployments")
+		}
+		taskQueue, wait := core.WorkerPool(ctx, jobs)
+		wipeoutApiDeployments(ctx, client, taskQueue, project+"/apis/-")
+		wait()
+	}
+	{
+		if Verbose {
+			log.Infof(ctx, "Deleting apis")
+		}
+		taskQueue, wait := core.WorkerPool(ctx, jobs)
+		wipeoutApis(ctx, client, taskQueue, project)
+		wait()
+	}
+	if Verbose {
+		log.Infof(ctx, "Wipeout complete with err=%+v", _err)
+	}
+	return _err
+}
+
+type ResourceKind int
+
+const (
+	API ResourceKind = iota
+	Version
+	Spec
+	Deployment
+	Artifact
+)
+
+type deleteResourceTask struct {
```

Comment:
These seems like a duplicate of `deleteTask` from the `registry delete` command. In #445 I suggested creating some reusable types for common tasks to avoid reinventing the wheel each time we want to do simple operations like creating/updating/deleting resources. Maybe now's the time?

The tasks can also be simpler by making them type-specific, which eliminates the "this should never happen" situation where we encounter an "unknown resource type" in these generic-esque tasks.

Reply:
partly addressed in #639